File: lib/dog/runtime/future.rb
#
# Copyright 2011 by Salman Ahmad (salman@salmanahmad.com).
# All rights reserved.
#
# Permission is granted for use, copying, modification, distribution,
# and distribution of modified versions of this work as long as the
# above copyright notice is included.
#

module Dog
  class Future < DatabaseObject
    collection "futures"
    
    attr_accessor :_id
    attr_accessor :value_id
    attr_accessor :value
    attr_accessor :queue
    attr_accessor :blocking_tracks
    attr_accessor :broadcast_tracks
    attr_accessor :handlers
    
    def initialize(id = nil, value = nil)
      self.value_id = id
      self.value = value
      self.queue = []
      self.blocking_tracks = []
      self.broadcast_tracks = []
      self.handlers = []
    end
    
    def self.remove_broadcast_track_from_all(track_id)
      self.update({
        "broadcast_tracks" => track_id
      }, { 
        "$pull" => { 
          "broadcast_tracks" => track_id 
        } 
      }, {
        :multi => true
      })
    end
    
    def to_hash
      blocking_tracks = self.blocking_tracks.map do |item|
        if item.kind_of? Track then
          item.save
          item._id
        elsif item.kind_of? BSON::ObjectId then
          item
        else
          raise "An invalid object appeared in the tracks list for a future"
        end
      end
      
      broadcast_tracks = self.broadcast_tracks.map do |item|
        if item.kind_of? Track then
          item.save
          item._id
        elsif item.kind_of? BSON::ObjectId then
          item
        else
          raise "An invalid object appeared in the tracks list for a future"
        end
      end
      
      queue = self.queue.map do |value|
        value.to_hash
      end
      
      value = self.value
      if self.value.nil? then
        value = nil
      else
        value = value.to_hash
      end
      
      hash = {
        "value_id" => self.value_id,
        "value" => value,
        "queue" => queue,
        "blocking_tracks" => blocking_tracks,
        "broadcast_tracks" => broadcast_tracks,
        "handlers" => handlers
      }
      
      return hash
    end
    
    def self.from_hash(hash)
      object = super
      
      if object.value != nil then
        object.value = ::Dog::Value.from_hash(object.value)
      end
      
      object.queue = object.queue.map do |item|
        ::Dog::Value.from_hash(item)
      end
      
      return object
    end
  end
end

File: src/main/java/dog/lang/Future.java

/*
 *
 *  Copyright 2012 by Salman Ahmad (salman@salmanahmad.com).
 *  All rights reserved.
 *
 *  Permission is granted for use, copying, modification, distribution,
 *  and distribution of modified versions of this work as long as the
 *  above copyright notice is included.
 *
 */

package dog.lang;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.json.JSONObject;
import com.mongodb.DBObject;
import com.mongodb.BasicDBObject;
import com.mongodb.MongoClient;
import com.mongodb.DB;
import com.mongodb.DBCollection;
import org.bson.types.ObjectId;

public class Future extends DatabaseObject {
	
	ObjectId valueId = null;
	Value value = null;
	
	ArrayList<Value> queue = new ArrayList<Value>();
	int queueSize = 0;

	ArrayList<Object> blockingStackFrames = new ArrayList<Object>();
	ArrayList<Object> broadcastStackFrames = new ArrayList<Object>();
	ArrayList<String> handlers = new ArrayList<String>();

	public Future(Runtime runtime) {
		this.runtime = runtime;
	}

	public String collectionName() {
		return "futures";
	}

	public Map toMap() {
		return toMongo().toMap();
	}

	public JSONObject toJSON() {
		return null;
	}

	public DBObject toMongo() {
		BasicDBObject hash = new BasicDBObject();
		hash.put("_id", this.getId());
		
		hash.put("value_id", this.valueId);
		hash.put("value", this.value.toMongo());

		ArrayList<DBObject> queueList = new ArrayList<DBObject>();
		for(Value v : queue) {
			queueList.add(v.toMongo());
		}

		hash.put("queue", queueList);
		hash.put("queue_size", this.queueSize);

		ArrayList<ObjectId> list = null;

		list = new ArrayList<ObjectId>();
		for(Object frame : this.blockingStackFrames) {
			if(frame instanceof StackFrame) {
				StackFrame theFrame = (StackFrame)frame;

				theFrame.setRuntime(runtime);
				theFrame.save();
				list.add(theFrame.getId());
			} else if(frame instanceof ObjectId) {
				list.add((ObjectId)frame);
			}
		}

		hash.put("blocking_stack_frames", list);

		list = new ArrayList<ObjectId>();
		for(Object frame : this.broadcastStackFrames) {
			if(frame instanceof StackFrame) {
				StackFrame theFrame = (StackFrame)frame;

				theFrame.setRuntime(runtime);
				theFrame.save();
				list.add(theFrame.getId());
			} else if(frame instanceof ObjectId) {
				list.add((ObjectId)frame);
			}
		}

		hash.put("broadcast_stack_frames", list);
		hash.put("handlers", this.handlers);

		return hash;
	}

	public void fromMap(Map map, Resolver resolver) {
		this.fromMongo(new BasicDBObject(map), resolver);
	}

	public void fromJSON(JSONObject json, Resolver resolver) {
		
	}

	public void fromMongo(DBObject bson, Resolver resolver) {
		this.id = (ObjectId)bson.get("_id");

		this.valueId = (ObjectId)bson.get("value_id");
		this.value = Value.createFromMongo((DBObject)bson.get("value"), resolver);

		for(Object o : (List)bson.get("queue")) {
			this.queue.add(Value.createFromMongo((DBObject)o, resolver));
		}
		this.queueSize = (Integer)bson.get("queue_size");

		for(Object object : (List)bson.get("blocking_stack_frames")) {
			this.blockingStackFrames.add(object);
		}

		for(Object object : (List)bson.get("broadcast_stack_frames")) {
			this.blockingStackFrames.add(object);
		}

		for(Object object : (List)bson.get("handlers")) {
			this.handlers.add((String)object);
		}


	}


}
